src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1789,6 +1954,120 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+  do
+  {
+    Assert(!curr.isNull());
+    success = false;
+    if (curr != zero && sindex < n1.size())
+    {
+      unsigned sindex_use = dir == 1 ? sindex : ((n1.size() - 1) - sindex);
+      if (n1[sindex_use].isConst())
+      {
+        // could strip part of a constant
+        Node lowerBound = getConstantArithBound(curr);
+        if (!lowerBound.isNull())
+        {
+          Assert(lowerBound.isConst());
+          Rational lbr = lowerBound.getConst<Rational>();
+          if (lbr.sgn() > 0)
+          {
+            Assert(checkEntailArith(curr));
+            CVC4::String s = n1[sindex_use].getConst<String>();
+            Node ncl =
+                NodeManager::currentNM()->mkConst(CVC4::Rational(s.size()));
+            Node next_s =
+                NodeManager::currentNM()->mkNode(kind::MINUS, lowerBound, ncl);
+            next_s = Rewriter::rewrite(next_s);
+            Assert(next_s.isConst());
+            // we can remove the entire constant
+            if (next_s.getConst<Rational>().sgn() >= 0)
+            {
+              curr = Rewriter::rewrite(
+                  NodeManager::currentNM()->mkNode(kind::MINUS, curr, ncl));
+              success = true;
+              sindex++;
+            }
+            else
+            {
+              // we can remove part of the constant
+              // lower bound minus the length of a concrete string is negative,
+              // hence lowerBound cannot be larger than long max
+              Assert(lbr < Rational(LONG_MAX));
+              curr = Rewriter::rewrite(NodeManager::currentNM()->mkNode(
+                  kind::MINUS, curr, lowerBound));
+              unsigned lbsize = lbr.getNumerator().toUnsignedInt();
+              Assert(lbsize < s.size());
+              if (dir == 1)
+              {
+                // strip partially from the front
+                nr.push_back(
+                    NodeManager::currentNM()->mkConst(s.prefix(lbsize)));
+                n1[sindex_use] = NodeManager::currentNM()->mkConst(
+                    s.suffix(s.size() - lbsize));
+              }
+              else
+              {
+                // strip partially from the back
+                nr.push_back(
+                    NodeManager::currentNM()->mkConst(s.suffix(lbsize)));
+                n1[sindex_use] = NodeManager::currentNM()->mkConst(
+                    s.prefix(s.size() - lbsize));
+              }
+              ret = true;
+            }
+            Assert(checkEntailArith(curr));
+          }
+          else
+          {
+            // we cannot remove the constant
+          }
+        }
+      }
+      else
+      {
+        Node next_s = NodeManager::currentNM()->mkNode(
+            kind::MINUS,
+            curr,
+            NodeManager::currentNM()->mkNode(kind::STRING_LENGTH,
+                                             n1[sindex_use]));
+        next_s = Rewriter::rewrite(next_s);
+        if (checkEntailArith(next_s))
```

Comment:
Additionally, we could check whether the lower bound of `curr` is larger than the upper bound of the length of `n1[sindex_use]` (using `getConstantArithBound`).

Reply:
I think this will also be covered by the generalization to abstract interpretation within checkEntailArith.
Basically, we will make checkEntailArith as sophisticated as possible, the internal details should be handled automatically, we don't need to argue special cases at the call locations to checkEntailArith, i.e. just passing curr-str.len(n1[sindex_use]) suffices, internally checkEntailArith will do what you mention.

Reply:
That makes sense.